### ISSUE TITLE
[⚠️ BREAKING CHANGE] - Embedding the public dashboard

### ISSUE THREAD

**Issue by alexjustesen** [alexjustesen/speedtest-tracker#1024]
## ⚠️ Breaking change

For those embedding their Speedtest Tracker public dashboards as of `0.14.2` this will be broken and stop functioning. TL;DR: Not secure, don't like that.

## 📜 Description

As of `v0.14.1` embedding the public dashboard into something like Home Assistant's dashboard or using an iframe was broken because of CORS and requiring a CSRF token to validate the request.

`0.14.2-beta` series tested removing CSRF protections on the dashboard. During exploration of this solution it was discovered that in Livewire, the package used to provide reactivity, there currently isn't a good way of disabling this on a specific route. 

The [solution](https://github.com/alexjustesen/speedtest-tracker/pull/1019) was to disable CSRF protection on _**ALL**_ `livewire/*` requests which could open an attack surface for dashboards exposed to the internet. **This isn't an acceptable solution IMO.**

### 🔗 Past linked issues

- #411
- #752

## 🤔 Proposed solution

Develop a new dashboard that is specifically for embedding into websites or other dashboards like Home Assistant. The dashboard should improve upon the current performance issues and not compromise security for features.

### 🛣️ Possible solutions

- If Livewire releases a 1st party easy way of disabling csrf on read-only requests, I'll used that https://github.com/livewire/livewire/discussions/7563
- Build a new dashboard that leverages an API instead of Livewire to update charts and metrics

## 🙋‍♂️ FAQs

- **What's the risk in just disabling csrf?**: [Cross Site Request Forgery (CSRFC)](https://owasp.org/www-community/attacks/csrf) provides protection against requests and actions that don't originate from the authenticated user or the application.
- **When do you plan on bringing back embeddable dashboards?**: Once I'm through fixing data issues in the next major release (`v0.15.0`) I'll be pivoting back to developing a secure and performant solution.

👇 I'll be tracking all research and updates in the comments below. Feel free to ask any questions or provide input.

**Comment by alexjustesen** [comment]
#1027 removes code related to embedding the public dashboard.

**Comment by alexjustesen** [comment]
Design idea, the dashboard should be big and bold so KPI's, charts and data are easily readable.

![image](https://github.com/alexjustesen/speedtest-tracker/assets/1144087/40e493d8-9b82-46df-a4cf-880be1f4dd6e)
_Tempest weather station dashboard_

**Comment by alexjustesen** [comment]
> any news here?

None yet, DQ issues need to be resolved first.

**Comment by fischerphilipp** [comment]
Not sure if this is news, but with the latest Home Assistant release (2024.4) I can simply embed the public dashboard via the new "Webpage" dashboard type. I just tried it and it works perfectly

**Comment by ZoXx** [comment]
iobroker hasnt this feature.
Need still public dashboard site for seeing it.

**Comment by ZoXx** [comment]
Any news @alexjustesen ? 🙂

**Comment by alexjustesen** [comment]
Explore using https://wire-elements.dev/blog/embed-livewire-components-using-wire-extender

**Comment by svenvg93** [comment]
@alexjustesen Could we replace this with an API endpoint— such as /results/latest or one returns only human-readable data? For example:

```
{
    "data": {
        "id": 563,
        "ping_human": "11 ms",
        "download_bits_human": "96.13 Mbps",
        "upload_bits_human": "28.14 Mbps",
        "created_at": "2025-06-03 11:59:26",
        "updated_at": "2025-06-03 11:59:37"
    },
    "message": "ok"
}
```

**Comment by ck99** [comment]
> > One possible approach to this would be to serve static json instead of dynamic. Ie. This content only changes after a speed test is run, and could therefore be generated/refreshed at that point.
> 
> It will always show the latest result. If you query it after a test it will be the last one. Normally we don't store/cache the response just query the DB

Yes, I understand - but my suggestion is that such a cached json solution would only need to be updated after a new test. Just like you store the result in the DB at that point, you could also append the data to a static file that is available directly from static assets.  Hooking this into the handling of a completed test, for example. 

For retrospective results, you could have a one time maintenance task that rebuilds the static file for all historic results. 

For aggregations etc, the static files could be segmented by day/week/month/year etc

My main point is that the content is not so dynamic, and also not very complicated. It could be served as flat static assets for consumption by external services, rather than requiring "online" support in the backend
